### ISSUE TITLE
Add functional Flask endpoints

### ISSUE THREAD

**Issue by hahahannes** [rucio/rucio#1817]
Motivation
----------
The Flask API has problems with dealing exceptions when using JSON streaming. This should be fixed as it is essential for handling lot of data.

Modification
------------
See [documentation](http://flask.pocoo.org/docs/1.0/patterns/streaming/?highlight=stream) for implementing JSON streaming to change the Flask API.
Investigate how to deal with exceptions.

**Comment by hahahannes** [comment]
It seems that it is not possible to catch exceptions inside the streaming.
To reproduce, run this minimal flask server. The exception gets raised on flask side but does not get catched. An internal server error is generated instead of the 'error' string.

```python
from flask import Flask, Response, jsonify

app = Flask(__name__)

@app.route("/")
def hello():
    def generate(): 
        for i in range(0, 100):   
            raise Exception
            yield '1' + '\n'    
    
    try:
        return Response(generate())
    except Exception as error:
        return jsonify({'error': error})

if __name__ == '__main__':
    app.run(passthrough_errors=True)
```

```
FLASK_APP=file.py flask run
```

One solution would be to rework the logic and check for possible problems like non existing account before querying the database. 
Another solution would be to call the same function inside a for loop but with a break after one loop, where we check for exceptions. Afterward the streaming can start if there was no exception catched.

**Comment by bari12** [comment]
To me the accurate path looks like the one where we change the logic (and checking) of the APIs which use streaming. Thus we do all the conflict checking first and then just iterate the stream.

Conceptually, the error handling we do right now with streaming is wrong and possibly leads to malformed replies. Because the HTTP return code is calculated (and sent!) at the header anyway. So let's say this is a `200`, whatever you stream afterwards (and throws an exception) is not properly reflected in the error code.

But this will take time, we probably have to more carefully select which APIs we offer as a streaming API (list-contents, list-replicas, ...) and not be so generous like we are now with them.

**Comment by vingar** [comment]
Worth mentioning here to think about proper support for HTTP Chunked Encoding.

**Comment by bziemons** [comment]
@bari12 *after* my current tasks I would like to take care of this. I debugged a little on the issue where there is only a generic internal server error transmitted to the client and what you pointed out above is most probably the cause of the problem.

**Comment by bari12** [comment]
👍 much appreciated. We need to plan and approach this carefully though. Especially with list_replicas it won't be easy.

**Comment by bziemons** [comment]
Just talked to Mario about it. He suggested the following tasks and their order:
- Move to Flask without breaking the API
- Add proper API versioning (`/v1/*`, `/v2/*`, etc.)
- Add help endpoint to retrieve API doc for a specific endpoint
- Repair streaming by using pagination (with an next page header). This point is a suggestion by me to actually solve our problems with errors coming from our current streaming endpoints. Using pagination in REST APIs is pretty common and the usual way to send an error to the client in the middle of retrieving information. We have to think about concurrent database updates in between retrieving pages though. Doing this correctly would actually save database session time and free them more quickly since we can retrieve a page on the server, free the session and we don't have to wait on the client to receive all the information.

**Comment by mlassnig** [comment]
pulling in @tbeerman for the discussion, since he already managed to port a lot of the API to flask

**Comment by mlassnig** [comment]
For the first bullet: i don't think it's possible to move to flask without breaking the API, since flask doesn't allow exceptions once you're already streaming. So the versioning should be step 1, no?

**Comment by bziemons** [comment]
> For the first bullet: i don't think it's possible to move to flask without breaking the API, since flask doesn't allow exceptions once you're already streaming. So the versioning should be step 1, no?

There is no way this is properly supported in web.py either. Once the header is sent, you can't change the status code. And I think we can mimic the current behavior, so that there is no breakage.

Edit: Also if we do the versioning first, I'd have to do part of the work twice - for all web.py and Flask endpoints.

**Comment by mlassnig** [comment]
Don't worry, we're doing it wrongly in web.py as well :-)

**Comment by tbeerman** [comment]
> There is no way this is properly supported in web.py either. Once the header is sent, you can't change the status code. And I think we can mimic the current behavior, so that there is no breakage.
> 
> Edit: Also if we do the versioning first, I'd have to do part of the work twice - for all web.py and Flask endpoints.

It's some time ago now so I don't completely recall the details but the main problem wasn't changing the header after it has been sent. In the current web.py way the rest app would call the core function. If there is something wrong the core function would raise an exception and the rest app would return the according error code. Otherwise it would just start streaming.
E.g., list_rses: if you give an expression if would call `parse_rse_expression`. If the expression is invalid it would raise `InvalidRSEExpression` which is then turned into the http error 400. Otherwise, it would just start yielding the results and web.py would return 200.

If I understand correctly how flask streaming works you have to use generators. You would have to call the core function inside the generator and then the http return code has already been set. So you would have to do those checks before you start the streaming, right?

As I said, that's already some time ago. Probably I'm just missing something obvious.

**Comment by bziemons** [comment]
> If I understand correctly how flask streaming works you have to use generators.

Just for my understanding: so it is the same as with web.py, correct?

> You would have to call the core function inside the generator and then the http return code has already been set.

I am not 100% familiar with how the code currently works, but isn't the core function the generator, which yields the results directly from the database? How is this different in web.py and Flask then?

> So you would have to do those checks before you start the streaming, right?

Which you would do in web.py and in Flask as good as possible. You cannot completely avoid errors while streaming - e.g. if the connection to the database is lost. And that exactly should be solved in the last task I suggested in https://github.com/rucio/rucio/issues/1817#issuecomment-680065138

**Comment by tbeerman** [comment]
> Just for my understanding: so it is the same as with web.py, correct?
>
Yes, you're right. They are both using generators. The difference is that in web.py you directly yield the results in the GET method of the corresponding class. In flask you have to pass a generator function to the Response object. Then this generator function is yielding the results.
In web.py
```
class RSEs(RucioController):
    @check_accept_header_wrapper(['application/x-json-stream'])
    def GET(self):
        header('Content-Type', 'application/x-json-stream')
        params = input()
        if 'expression' in params:
            try:
                for rse in parse_rse_expression(params['expression'], vo=ctx.env.get('vo')):
                    item = {'rse': rse}
                    yield render_json(**item) + '\n'
            except InvalidRSEExpression as error:
                raise generate_http_error(400, 'InvalidRSEExpression', error.args[0])
```
In flask it would be something like this:
```
class RSEs(MethodView):
    @check_accept_header_wrapper(['application/x-json-stream'])
    def get(self):
        def generate():
            for rse in parse_rse_expression(request.args['expression'], vo=ctx.env.get('vo')):
                item = {'rse': rse}
                yield render_json(**item) + '\n'
        return Response(stream_with_context(generate()))
```
I couldn't find out how to change the response as soon as you are in the generator function. But with the current core function you would only now if there are any exception as soon as you call them. But maybe there is a way to handle the exception inside the generator function an pass it to the response but I couldn't find it.

> Which you would do in web.py and in Flask as good as possible. You cannot completely avoid errors while streaming - e.g. if the connection to the database is lost. And that exactly should be solved in the last task I suggested in [#1817 (comment)](https://github.com/rucio/rucio/issues/1817#issuecomment-680065138)

I think the errors during streaming are not that problematic. As you said they can always occur. The thing is that we would need to change the core somehow so that we can do the checks before we start the streaming.

**Comment by bziemons** [comment]
> I think the errors during streaming are not that problematic.

Well for the clients they actually are, because right now there is no way to tell the client about the details.

> The thing is that we would need to change the core somehow so that we can do the checks before we start the streaming.

I think this is obsolete once there is actually a good error reporting even from the generator functions. Why should we worry about this then?

**Comment by tbeerman** [comment]
I'm not sure that I can follow anymore. You're first point was "Move to Flask without breaking the API". In that case there wouldn't be any changes to the actual streaming, yet. I'm a right? How would the actual error handling for all the errors that can occur before the streaming be done in that case?

**Comment by bziemons** [comment]
As promised and in color. Something along those lines:

```python
def try_stream(generator, successcb, failurecb):
    it = iter(generator)
    try:
        peek = next(it)
        return successcb(itertools.chain((peek,), it))
    except StopIteration:
        return successcb(iter(()))
    except Exception as error:
        return failurecb(error)


def response_from_error(error):
    if isinstance(error, InvalidRSEExpression):
        return generate_http_error_flask(400, 'InvalidRSEExpression', error.args[0])
    # elif other error cases
    else:
        return generate_http_error_flask(500, error.__class__.__name__, "No error message" if not error.args else error.args[0])


class RSEs(MethodView):
    @check_accept_header_wrapper(['application/x-json-stream'])
    def get(self):
        def generate():
            for rse in parse_rse_expression(request.args['expression'], vo=ctx.env.get('vo')):
                item = {'rse': rse}
                yield render_json(**item) + '\n'

        return try_stream(generate(), lambda it: Response(stream_with_context(it)), response_from_error)
```

**Comment by bari12** [comment]
Hi,

Just coming into this post-vacation :-)
I think the task list from above is good. Trying to replicate the wrong behaviour from web.py with flask at least allows us to switch to flask and still keep the same `/v1/ `endpoints.
Switching to a (paginated) v2 API will be a much bigger step and we need to map out how we want to handle backwards compatibility with this, but laying all the groundwork and switching to flask already is the right step. 👍

**Comment by bziemons** [comment]
Got something prepared for fixing flask endpoints #3975 and testing with Flask instead of web.py #3990 -- the first PR can be merged after #3995 and the second still has a task open.
